certbot-dns-bookmyname/certbot_dns_bookmyname/dns_bookmyname.py

```diff
@@ -0,0 +1,330 @@
+                except Error as e:
+                    raise errors.PluginError(
+                        'Error getting BookMyName domains: {0}'.format(e)
+                    )
+            except Error as e:
+                raise errors.PluginError(
+                    'Error logging into BookMyName: {0}'.format(e)
+                )
+
+    def get_registered_domain(self, domain):
+        """
+        Return the registered domain matching the provided domain if any
+
+        A registered domain matches the provided domain if the provided
+        domain ends with the registered domain.
+
+        This method will log in if needed.
+        """
+        self.login()
+        for registered_domain in self.domains_and_ids:
+            if domain.endswith(registered_domain):
+                return registered_domain
+        raise errors.PluginError(
+            'No registered BookMyName domain match for domain {0}'
+            .format(domain)
+        )
+
+    def get_registered_domain_id(self, domain):
+        """
+        Return the ID of the matching registered domain if any
+
+        The ID is needed to pass to the website when setting the zone
+        contents (whereas getting the zone contents takes the domain,
+        not ID. Go figure).
+
+        This method will log in if needed.
+        """
+        self.login()
+        registered_domain = self.get_registered_domain(domain)
+        return self.domains_and_ids[registered_domain]
+
+    def add_txt_record(self, domain, validation_name, validation):
+        """
+        Add a TXT record for the validation name and value
+
+        The record sets as low a TTL as BookMyName allows, which is 300.
+
+        The validation name has the full domain, but the record expects
+        only the prefix, without the registered domain at the end, so we
+        remove the suffix.
+
+        This method will log in if needed.
+        """
+        self.login()
+        try:
+            registered_domain = self.get_registered_domain(domain)
+            registered_domain_id = self.get_registered_domain_id(domain)
+            short_valid_name = re.sub(r"\."+registered_domain+"$", "", validation_name)
+            record = short_valid_name + ' 300 TXT "' + validation +'"'
+            zone_in = self.manager.get_domain_zone(registered_domain_id)
+            zonelines_raw = zone_in.split('\n')
+            zonelines = [' '.join(X.split()) for X in zonelines_raw]
+            zonelines.append(record)
+            zone_out = '\n'.join(zonelines)
+            self.manager.set_domain_zone(registered_domain_id, zone_out)
+        except Error as e:
+            raise errors.PluginError(
+                'Error adding BookMyName domain {0} TXT record {1} = {2}: {3}'
+                .format(registered_domain, short_valid_name, validation, e)
+            )
+
+    def del_txt_record(self, domain, validation_name, validation):
```

Comment:
This method and `add_txt_record` seems to include a lot of duplicate logic. Can we introduce some private methods to consolidate the common behaviors?

Reply:
Sure.

Reply:
Done